src/soc/ibm/power9/homer.c

```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+			/*
+			 * Need an 'isync' here to ensure that previous instructions
+			 * have completed before the code continues on. This is a type
+			 * of read-barrier.  Without this the processor can do
+			 * speculative reads of the HOMER data and you can actually
+			 * get stale data as part of the instructions that happen
+			 * afterwards. Another 'weak consistency' issue.
+			 */
+			asm volatile("isync" ::: "memory");
+
+			/* OCC must have processed the command */
+			break;
+		}
+
+		if (timeout_ms > 0) {
+			/* Delay before the next check */
+			int32_t sleep_ms = OCC_RSP_SAMPLE_TIME_MS;
+			if (timeout_ms < sleep_ms)
+				sleep_ms = timeout_ms;
+
+			wait_ms(sleep_ms, false);
+			timeout_ms -= sleep_ms;
+		} else {
+			/* Time expired */
+			die("Timed out while waiting for OCC response\n");
+		}
+	}
+}
+
+static bool parse_occ_response(struct homer_st *homer, uint8_t occ_cmd,
+			       uint8_t *status, uint8_t *seq_num,
+			       uint8_t *response, uint32_t *response_len)
+{
+	uint16_t index = 0;
+	uint16_t data_len = 0;
+	uint16_t checksum = 0;
+	uint16_t i = 0;
+
+	const uint8_t *rsp_buf = &homer->occ_host_area[OCC_RSP_ADDR];
+
+	*seq_num = rsp_buf[index++];
+	index += 1; /* command */
+	*status = rsp_buf[index++];
+
+	data_len = *(uint16_t *)&rsp_buf[index];
+	index += 2;
+
+	if (data_len > 0) {
+		uint16_t copy_size = data_len;
+		if (copy_size > *response_len)
+			copy_size = *response_len;
+
+		memcpy(response, &rsp_buf[index], copy_size);
+		*response_len = copy_size;
+
+		index += data_len;
+	}
+
+	for (i = 0; i < index; ++i)
+		checksum += rsp_buf[i];
+
+	if (checksum != *(uint16_t *)&rsp_buf[index]) {
+		printk(BIOS_WARNING, "OCC response for 0x%02x has invalid checksum\n",
+		       occ_cmd);
+		return false;
+	}
+
+	if (*status != OCC_RC_SUCCESS) {
+		printk(BIOS_WARNING, "0x%02x OCC command failed with an error code: 0x%02x\n",
+		       occ_cmd, *status);
+		return false;
+	}
+
+	return true;
+}
+
+static bool write_occ_cmd(struct homer_st *homer, uint8_t occ_cmd,
+			  const uint8_t *data, uint16_t data_len,
+			  uint8_t *response, uint32_t *response_len)
+{
+	static uint8_t cmd_seq_num;
+
+	uint8_t status = 0;
+	uint8_t rsp_seq_num = 0;
+
+	++cmd_seq_num;
+	/* Do not use 0 for sequence number */
+	if (cmd_seq_num == 0)
+		++cmd_seq_num;
+
+	build_occ_cmd(homer, occ_cmd, cmd_seq_num, data, data_len);
+	/* Sender: HTMGT; command: Command Write Attention */
+	write_occ_command(0x1001000000000000);
+
+	/* Wait for OCC to process command and send response (timeout is the
+	 * same for all commands) */
+	wait_for_occ_response(homer, 20, cmd_seq_num);
+
+	if (!parse_occ_response(homer, occ_cmd, &status, &rsp_seq_num, response,
+				response_len)) {
+		printk(BIOS_WARNING, "Received OCC response:\n");
+		hexdump(response, *response_len);
+		printk(BIOS_WARNING, "Failed to parse OCC response\n");
+		return false;
+	}
+
+	/* Statuses of 0xE0-EF are reserved for OCC exceptions */
+	if ((status & 0xF0) == 0xE0)
```

Comment:
Is it even possible? parse_occ_response() checks for OCC_RC_SUCCESS and returns false, so previous `if` will be taken in that case.